pkg/infradb/utils.go

```diff
@@ -0,0 +1,294 @@
+
+	var deferErr error
+	var preSubscriber *actionbus.Subscriber
+
+	defer func() {
+		if deferErr != nil {
+			log.Println("startReplayProcedure(): The replay procedure has failed")
+			log.Println("startReplayProcedure(): unblocking the TaskManager to continue")
+			taskmanager.TaskMan.ReplayFinished()
+		}
+	}()
+
+	preSubscribers := actionbus.ABus.GetSubscribers("preReplay")
+
+	for _, preSub := range preSubscribers {
+		if preSub.Name == componentName {
+			preSubscriber = preSub
+			break
+		}
+	}
+
+	if preSubscriber == nil {
+		deferErr = fmt.Errorf("no pre-replay subscriber for %s", componentName)
+		log.Printf("startReplayProcedure(): Error %+v\n", deferErr)
+		return
+	}
+
+	// Notify the preReplay subscriber
+	actionData := actionbus.NewActionData()
+	deferErr = actionbus.ABus.Publish(actionData, preSubscriber)
+	if deferErr != nil {
+		log.Printf("startReplayProcedure(): Error %+v\n", deferErr)
+		return
+	}
+
+	// Waiting for the pre-replay procedure to finish
+	deferErr = <-actionData.ErrCh
+	close(actionData.ErrCh)
+
+	if deferErr != nil {
+		log.Printf("startReplayProcedure(): Error %+v\n", deferErr)
+		return
+	}
+
+	log.Printf("startReplayProcedure(): Component %s has successfully executed pre-replay steps", componentName)
+
+	objectTypesToReplay := getObjectTypesToReplay(componentName)
+
+	objectsToReplay, subsForReplay, err := gatherObjectsAndSubsToReplay(componentName, objectTypesToReplay)
+
+	if err != nil {
+		log.Printf("startReplayProcedure(): Error %+v\n", err)
+		deferErr = err
+		return
+	}
+
+	// Releasing the lock as all the operations in the DB has finished
+	globalLock.Unlock()
+
+	// Notify task manager to continue processing tasks as
+	// the replay of objects in the DB has finished
+	taskmanager.TaskMan.ReplayFinished()
+
+	createReplayTasks(objectsToReplay, subsForReplay)
+}
+
+// getObjectTypesToReplay collects all the types of object to be replayed
+// which are related to the component that called the replay.
+func getObjectTypesToReplay(componentName string) []string {
+	objectTypesToReplay := []string{}
+
+	bpSubs := eventbus.EBus.GetSubscribers("bridge-port")
+	sviSubs := eventbus.EBus.GetSubscribers("svi")
+	lbSubs := eventbus.EBus.GetSubscribers("logical-bridge")
+	vrfSubs := eventbus.EBus.GetSubscribers("vrf")
+
+	for _, vrfSub := range vrfSubs {
+		if vrfSub.Name == componentName {
+			objectTypesToReplay = append(objectTypesToReplay, "vrf")
+			break
+		}
+	}
+
+	for _, lbSub := range lbSubs {
+		if lbSub.Name == componentName {
+			objectTypesToReplay = append(objectTypesToReplay, "logical-bridge")
+			break
+		}
+	}
+
+	for _, sviSub := range sviSubs {
+		if sviSub.Name == componentName {
+			objectTypesToReplay = append(objectTypesToReplay, "svi")
+			break
+		}
+	}
+
+	for _, bpSub := range bpSubs {
+		if bpSub.Name == componentName {
+			objectTypesToReplay = append(objectTypesToReplay, "bridge-port")
+			break
+		}
+	}
+
+	return objectTypesToReplay
+}
+
+// nolint: funlen, gocognit
```

Comment:
I'd prefer to fix linter warnings, rather than suppressing them.

Reply:
Is not easy to make this function smaller. It will create confusion and it will become error prone. I think I will ass this in the context of the restructuring effort. I can open an issue for that.

Reply:
why? They look similar, you can move the code into the corresponding function/functions

Reply:
I can probably create 4 functions that each one of those handle the different types of cases/objects (vrf,lb,bp,svi) in the switch statement. Is that ok ? I cann not create one function to handle all the cases as we are talking about different structs to be handled by the same function and that is really difficult to implement in golang except if you have something else to propose. But 4 functions one per case I can create without a problem. WDYT ?

Reply:
Generics(less preferable)/interfaces(more preferable)?
If `infradb.client.` could return not only the concrete types, but also some common interface(slice of such interface instances) with required operations like `GetName`, `IsReplayNeeded`, `GetStatus` then I think it would be possible

Reply:
Yes I am not sure how to change this right now.  I think we should think this through better when we have more time. Let's add this to the refactoring topic. I can open a github issue to track this if that is alright. I will suggest to park this for now until we have a good design that will address the refactorng topic in general